### PR TITLE
ULP-502: LinkedIn API v2 (based on #63)

### PR DESCRIPTION
**Pull request by sebadoom**
This updates the strategy to LinkedIn's new API v2: https://docs.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/migration-faq?context=linkedin/consumer/context

Note that the "r_basicprofile" fields for the user profile have not been validated, this will require access to LinkedIn's Partner Profile. Setting this scope without being part of the program will cause the scope to be ignored without errors. Note that if this scope is used without being part of the program and "r_liteprofile" is not specified, "r_basicprofile" will be ignored and no scope for profile access will be available, causing LinkedIn to return:

```
{ 
  statusCode: 403,
  data: '{"serviceErrorCode":100,"message":"Not enough permissions to access: GET /me","status":403}' 
}
```

### CODE REVIEW

File: example/server.js [hunk 1]
```diff
@@ -4,8 +4,9 @@ var express = require('express')
 
 // API Access link for creating client ID and secret:
 // https://www.linkedin.com/secure/developer
-var LINKEDIN_CLIENT_ID = "50p3e3wo29te";
-var LINKEDIN_CLIENT_SECRET = "CUFBLLzfpU24oQ3B";
```
Comment (alejofernandez) [inline example/server.js]: 😱😱😱

File: lib/oauth2.js [hunk 2]
```diff
@@ -1,184 +1,163 @@
...
   }
 
-  return fields.join(',');
+  profile._emailRaw = body;
+  profile._emailJson = json;
+
+  return profile;
 }
 
+var liteProfileUrl = 'https://api.linkedin.com/v2/me?projection=(' +
```
Comment (glena) [inline lib/oauth2.js]: can we move this to the top?

File: lib/oauth2.js [hunk 3]
```diff
@@ -1,184 +1,163 @@
...
+  'id,' +
+  'firstName,' +
+  'lastName,' +
+  'maidenName,' +
+  'profilePicture(displayImage~:playableStreams)' +
+  ')';
+
+var basicProfileUrl = 'https://api.linkedin.com/v2/me?projection=(' +
```
Comment (glena) [inline lib/oauth2.js]: same

File: example/server.js (hunk 1 above)
Comment (sebadoom) [reply example/server.js]: :scream:

**Comment by jefftmills** [comment]
Do you anticipate merging and releasing this before LinkedIn's March 1 deadline? I'd love to just update this to a new version vs. building my own solution.

Also, we're in their Partner program so I'm glad to see you kept the option of asking for the basic profile.

**Comment by sebadoom** [comment]
> Do you anticipate merging and releasing this before LinkedIn's March 1 deadline? I'd love to just update this to a new version vs. building my own solution.
> 
> Also, we're in their Partner program so I'm glad to see you kept the option of asking for the basic profile.

We expect to have a release soon. It should be out before the March 1st deadline.

**Comment by jefftmills** [comment]
I don't know if you noticed, but 2 days ago LinkedIn deprecated most of the fields in r_basicprofile. I guess the only impact here would be the list of fields to ask for when r_basicprofile is passed in, but I thought I'd pass this along.

https://docs.microsoft.com/en-us/linkedin/shared/references/v2/profile/basic-profile?context=linkedin/consumer/context

**Comment by Exide** [comment]
March 1st is in less than 3 days. Anything I can do to help get this merged? I really want to avoid forking/use a different package.

**Comment by andreyrd** [comment]
My team is going with the fork https://github.com/sokratisvidros/passport-linkedin-oauth2 considering we now only have 2 days left to do any sort of testing

**Comment by jybleau** [comment]
> My team is going with the fork https://github.com/sokratisvidros/passport-linkedin-oauth2 considering we now only have 2 days left to do any sort of testing

We already did so. Working fine with `r_liteprofile` scope, etc. No big deal.

But still waiting for this repo though... Let's hope.

**Pull request merged**